### PR TITLE
Propose a new challenge about pointer arithmetic ops

### PR DESCRIPTION
**Pull request by celinval**
In this challenge, we want to look at the safe usage of pointer arithmetic operations.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: doc/src/challenges/0003-pointer-arithmentic.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# Challenge 3: Verifying Raw Pointer Arithmetic Operations
+
+- **Status:** Open
+- **Solution:**
+- **Tracking Issue:** <https://github.com/model-checking/verify-rust-std/issues/21>
+- **Start date:** 24/06/24
+- **End date:** 24/12/10
```
Comment (tautschnig) [inline doc/src/challenges/0003-pointer-arithmentic.md:7]: I should have said so when reviewing the template, perhaps you can fix it as part of this PR: can we please use ISO-8601 date format (YYYY-MM-DD)? There are a lot of 24s here, and one has to read both dates to get an idea which will be the months and the day.

File: doc/src/challenges/0003-pointer-arithmentic.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+-------------------
+
+
+## Goal
+
+The goal of this challenge is to verify safety of code that relies on raw pointer arithmetic, and eventual
+raw pointer access.
```
Comment (tautschnig) [inline doc/src/challenges/0003-pointer-arithmentic.md:15]: I think this needs more precise scoping: which "code" that "relies on raw pointer arithmetic" do you actually want to cover?

File: doc/src/challenges/0003-pointer-arithmentic.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+However, bugs in these abstractions may compromise entire applications, potentially becoming a security concern.
+See [CVE-2018-1000810](https://www.cvedetails.com/cve/CVE-2018-1000810/) for an example of an issue with an
+optimization of `str::repeat`.
+
+These safe abstractions are great candidates for software verification.
+They are critical for Rust applications safety, and they are modular by design.
+
+## Description
+
+Rust provide different options for pointer arithmetic, which have different semantics when it comes to safety.
```
Comment (tautschnig) [inline doc/src/challenges/0003-pointer-arithmentic.md:35]:
```suggestion
Rust provides different options for pointer arithmetic, which have different semantics when it comes to safety.
```

Comment (tautschnig) [reply doc/src/challenges/0003-pointer-arithmentic.md:35]: Also: what does it mean to "have different semantics when it comes to safety?"

File: doc/src/challenges/0003-pointer-arithmentic.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+For example, methods such as [`ptr::offset`](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset),
+[`ptr::add`](https://doc.rust-lang.org/std/primitive.pointer.html#method.add),
+and [`ptr::sub`](https://doc.rust-lang.org/std/primitive.pointer.html#method.sub)
+are unsafe, and one of their safety condition is that:
```
Comment (tautschnig) [inline doc/src/challenges/0003-pointer-arithmentic.md:39]:
```suggestion
are unsafe, and one of their safety conditions is that:
```

File: doc/src/challenges/0003-pointer-arithmentic.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+> - Both the starting and resulting pointer must be either in bounds or one byte past the end of the same allocated object.
+
+I.e., violating this safety condition triggers immediate UB.
+
+On the other hand, wrapping operations such as
+[`ptr::wrapping_offset`](https://doc.rust-lang.org/std/primitive.pointer.html#method.wrapping_offset),
+[`ptr::wrapping_add`](https://doc.rust-lang.org/std/primitive.pointer.html#method.wrapping_add),
+[`ptr::wrapping_sub`](https://doc.rust-lang.org/std/primitive.pointer.html#method.wrapping_sub),
+are safe, however, the resulting pointer must not be used to read or write other allocated objects.
```
Comment (tautschnig) [inline doc/src/challenges/0003-pointer-arithmentic.md:48]: Possibly stupid question: what is the resulting pointer good for if it cannot be used to read or write (other?) objects?

File: doc/src/challenges/0003-pointer-arithmentic.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+Thus, we would like to be able to verify usage of these different methods within the standard library
+to ensure they are safely employed,
+as well as provide function contract that can be used by other Rust crates to verify their own usage of these methods.
```
Comment (tautschnig) [inline doc/src/challenges/0003-pointer-arithmentic.md:52]:
```suggestion
as well as provide function contracts that can be used by other Rust crates to verify their own usage of these methods.
```

File: doc/src/challenges/0003-pointer-arithmentic.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+### Assumptions
+
+For this challenge, we do not require a full proof that is independent of the pointee type `T`.
+Instead, we require that the verification is done for the following pointee types:
+1. All integer type.
```
Comment (tautschnig) [inline doc/src/challenges/0003-pointer-arithmentic.md:58]:
```suggestion
1. All integer types.
```

Comment (tautschnig) [reply doc/src/challenges/0003-pointer-arithmentic.md:58]: Also: can you please enumerate them? I wouldn't be surprised if those change over time (e.g., adding types greater than 128 bits wide).

File: doc/src/challenges/0003-pointer-arithmentic.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+2. At least one `dyn Trait`.
+3. At least one slice.
+4. For unit type.
+5. At least one composite types with multiple non-ZST fields.
```
Comment (tautschnig) [inline doc/src/challenges/0003-pointer-arithmentic.md:62]:
```suggestion
5. At least one composite type with multiple non-ZST fields.
```

**Review by tautschnig — APPROVED** [review]

File: doc/src/challenges/0003-pointer-arithmentic.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+## Goal
+
+The goal of this challenge is to verify safety of code that relies on raw pointer arithmetic, and eventual
+raw pointer access.
+
+## Motivation
+
+Raw pointer arithmetic is a common operation employed in the implementation of highly optimized code,
+as well as containers with dynamic size.
+Examples of the former are `str::repeat`, `[u8]::is_asc_ii`,
```
Comment (zhassan-aws) [inline doc/src/challenges/0003-pointer-arithmentic.md:21]:
```suggestion
Examples of the former are `str::repeat`, `[u8]::is_ascii`,
```

File: doc/src/challenges/0003-pointer-arithmentic.md (hunk 1 above)
Comment (celinval) [reply doc/src/challenges/0003-pointer-arithmentic.md:7]: Do you want a separate PR for that?

File: doc/src/challenges/0003-pointer-arithmentic.md (hunk 5 above)
Comment (celinval) [reply doc/src/challenges/0003-pointer-arithmentic.md:48]: I've seen the standard library using the address to implement ZST iterator that stores 2 pointers (current and end). For ZST, the length becomes the offset between them in bytes.

Comment (tautschnig) [reply doc/src/challenges/0003-pointer-arithmentic.md:48]: How is that still correct in the event of a wrap-around? (I am not sure which part in all this uses `wrapping_*` operations, but either way the result does not seem correct *if a wrap-around happens*?!)

File: doc/src/challenges/0003-pointer-arithmentic.md (hunk 2 above)
Comment (tautschnig) [reply doc/src/challenges/0003-pointer-arithmentic.md:15]: Can you point me to where exactly this is happening? I read "usage of these different methods within the standard library" and it makes me think this challenge requires verification of the _entire_ standard library.

Comment (celinval) [reply doc/src/challenges/0003-pointer-arithmentic.md:15]: That's the success criteria where we list all the functions that must be annotated and verified.

File: doc/src/challenges/0003-pointer-arithmentic.md (hunk 5 above)
Comment (celinval) [reply doc/src/challenges/0003-pointer-arithmentic.md:48]: That's a user choice. From the language and library perspective, anything here is OK. Maybe you can raise your concerns with the library team.

File: doc/src/challenges/0003-pointer-arithmentic.md (hunk 1 above)
Comment (celinval) [reply doc/src/challenges/0003-pointer-arithmentic.md:7]: I'll create a separate PR

**Pull request merged**